pkg/hooks/README.md

```diff
@@ -0,0 +1,177 @@
+This can cause some performance issues.
+Also a lot of hooks just check if certain configuration is set and then exit early, without doing anything.
+For example the [oci-systemd-hook][] only executes if the command is `init` or `systemd`, otherwise it just exits.
+This means if we automatically enabled all hooks, every container would have to execute `oci-systemd-hook`, even if they don't run systemd inside of the container.
+Performance would also suffer if we exectuted each hook at each stage ([pre-start][], [post-start][], and [post-stop][]).
+
+## Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" are to be interpreted as described in [RFC 2119][rfc2119].
+
+## JSON Definition
+
+This package reads all [JSON][] files (ending with a `.json` extention) from a series of hook directories.
+For both `crio` and `podman`, the directory is set via `--hooks-dir-path` (FIXME: undocumented option for both `crio` and `podman`?).
+
+For CRI-O, if `--hooks-dir-path` is not used to override the default directory, hook JSON is also read from `/etc/containers/oci/hooks.d/*.json`.
+If the same file is in both directories, the one in `/etc/containers/oci/hooks.d` takes precedence.
+
+Each JSON file should contain an object with the following properties:
+
+### 1.0.0 Hook Schema
+
+* **`version`** (REQUIRED, string) Sets the hook-definition version.
+    For this schema version, the value MUST be 1.0.0.
+* **`hook`** (REQUIRED, object) The hook to inject, with the [hook-entry schema][spec-hooks] defined by the 1.0.1 OCI Runtime Specification.
+* **`when`** (REQUIRED, object) Conditions under which the hook is injected.
+    The following properties can be specified:
+
+    * **`always`** (OPTIONAL, boolean) If set `true`, the hook MUST always be injected.
+    * **`annotations`** (OPTIONAL, object) The hook MUST be injected if an `annotations` key/value pair both match a key/value pair from the [configured annotations][spec-annotations].
+        Both keys and values MUST be [POSIX extended regular expressions][POSIX-ERE].
+    * **`commands`** (OPTIONAL, array of strings) The hook MUST be injected if the configured [`process.args[0]`][spec-process] matches an entry.
+        Entries MUST be [POSIX extended regular expressions][POSIX-ERE].
+    * **`hasBindMounts`** (OPTIONAL, boolean) The hook MUST be injected if `hasBindMounts` is true and the container is configured to bind-mount host directories into the container.
+        libpod does not consider this condition when matching hooks.
+* **`stages`** (REQUIRED, array of strings) Stages when the hook MUST be injected.
+    Entries MUST be chosen from the 1.0.1 OCI Runtime Specification [hook stages][spec-hooks].
+
+The matching properties in `when` are orthogonal, and the hook is injected if *any* of those properties match.
```

Comment:
I tgink I'll change this from "any" (i.e. or) to "all" (i.e. and) for 1.0.0.  That lets you more narrowly target injection.  Folks can recover "or" semantics by using multiple  JSON files.  Thoughts?

Reply:
Yes all sounds good.  Since we now support multiple.  That way a vendor could say run my hook if you have bind mounts and annotations.